src/reduce.jl

```diff
@@ -0,0 +1,130 @@
+export @groupreduce, Reduction
+
+module Reduction
+    const thread = Val(:thread)
+    const warp = Val(:warp)
+end
+
+"""
+    @groupreduce op val neutral algo [groupsize]
+
+Perform group reduction of `val` using `op`.
+
+# Arguments
+
+- `algo` specifies which reduction algorithm to use:
+    - `Reduction.thread`:
+        Perform thread group reduction (requires `groupsize * sizeof(T)` bytes of shared memory).
+        Available accross all backends.
+    - `Reduction.warp`:
+        Perform warp group reduction (requires `32 * sizeof(T)` bytes of shared memory).
+        Potentially faster, since requires fewer writes to shared memory.
+        To query if backend supports warp reduction, use `supports_warp_reduction(backend)`.
```

Comment:
Why is that needed? Shouldn't the backend go and use warp reductions if it can?

Reply:
I'm now doing an auto-selection of the algorithm based on device function `__supports_warp_reduction()`.